File: problems/cpp/lowest-common-ancestor-of-a-binary-search-tree.cc
// 26 ms, 09/12/2017
// Time  : O(h)
// Space : O(h)
// Note  : This problem guarantees that nodes p and q exist in the tree and
//         also that the tree has all distinct keys
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        if (root == p || root == q)
            return root;
        if (p->val < root->val && q->val < root->val)
            return lowestCommonAncestor(root->left, p, q);
        if (p->val > root->val && q->val > root->val)
            return lowestCommonAncestor(root->right, p, q);
        return root;
    }
};

File: code/python/lowest-common-ancestor-of-a-binary-search-tree.py
# 102 ms, 02/07/2018
# Time  : O(h)
# Space : O(h)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if not root or not p or not q:
            return root

        def lca_recur(root, s, l):
            return (lca_recur(root.right, s, l) if root.val < s.val
                    else lca_recur(root.left, s, l) if root.val > l.val
                    else root)
        # Make sure p has the smaller key
        if p.val > q.val:
            p, q = q, p
        return lca_recur(root, p, q)
